[util] Added Python script to generate a status report of the DIFs.

timothytrippel opened the pull request:
This utility can be used to generate the table(s) shown below that:
- Cross reference commit dates between hardware and DIF software to indicate which DIFs *may* need updating.
- Determine who the main authors are for both hardware IP and DIF software to track down implementation details of each.
- Indicate if a DIF API is currently incomplete, based on the function definitions in the DIF header file.

Usage: check_dif_statuses [-h] [--show-unimplemented] repo_top

Example DIF Status Table:
<img width="1424" alt="dif_statuses_ex_table" src="https://user-images.githubusercontent.com/5633066/127069122-6098f062-6071-4f44-86aa-8d5dcbd9c304.png">

Example Unimplemented DIF Functions Table:
<img width="573" alt="unimplemented_funcs_ex_table" src="https://user-images.githubusercontent.com/5633066/127069135-908a940c-2993-4884-81ae-523dee2aa424.png">

Signed-off-by: Timothy Trippel <ttrippel@google.com>

timothytrippel commented:
@tjaychen @cfrantz @mcy please review

silvestrst commented:
Hey @timothytrippel , I have noticed couple small mistakes. `DIF_UART` was actually written by me, and I think (but might be wrong) that `DIF_GPIO` was written by @alphan (but worth double checking with him).

alphan commented:
> Hey @timothytrippel , I have noticed couple small mistakes. `DIF_UART` was actually written by me, and I think (but might be wrong) that `DIF_GPIO` was written by @alphan (but worth double checking with him).

Yes, I guess this is due to the renames/updates that @mcy did (thanks). Didn't check which metric you are using to figure out the author but you may want to list two names (maybe github usernames to keep it shorter) if it won't be a big change, may be helpful when folks are OOO or move on to something else. Not a big deal though.

timothytrippel commented:
> Hey @timothytrippel , I have noticed couple small mistakes. `DIF_UART` was actually written by me, and I think (but might be wrong) that `DIF_GPIO` was written by @alphan (but worth double checking with him).
> 
> P.S.
> 
> Pretty cool tool, thanks! :)

@silvestrst  My apologies! The current way the tool computes the main "HW/DIF Authors" is based on most lines of code (LOC) committed, where  LOC = additions + deletions. As @alphan noted, the reason @mcy is showing up as the main author is because he must have done some refactoring. 

Another thing @alphan noted was that I should be careful to makes sure that "author" is not confused with "committer".

Additionally, @imphil good point, I will see if I can work that into the tool too.

---TL;DR---
To address all of the above, I will:
- [ ] alter the way main authors are determined to list the top two authors if there are two authors that contribute more than 5-10% of the LOC. (This should at least give an interested party a general point-of-contact for a *potential* DIF owner.)
- [ ] switch author names to github usernames for brevity
- [ ] add DIF lifecycle state to table
- [ ]  look into if it is possible to indicate author's status as active contributor to the project (not sure if the backend tools I am using expose such information or if there is a way to retrieve it.)

timothytrippel commented:
In addition to completing the above checklist items, I also:

- verified authors, not committers, are indeed listed (the python module I am using, git-fame, uses the `git blame` command under the hood which make this distinction).
- removed the "completed?" column, as this information is redundant given the color coding scheme/legend
- added multi-line cells to table for width reduction and better displaying in the terminal

One thing to note: when I was grabbing the DIF lifecycle states from the <ip>.prj.hjson files, if the IP had multiple revisions, I simply grabbed the latest.

cfrantz commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,307 @@
+        "repo_top",
+        help=
+        """Relative path to where the OpenTitan repository is checked out.""")
+    parser.add_argument(
+        "--show-unimplemented",
+        action='store_true',
+        help="""Show unimplemented functions for each incomplete DIF.""")
```
Just to stay honest about this: The notion of completeness of a DIF based on implemented functions is very naive.  It assumes the header file has a complete description of the API for that DIF and assumes that any implemented function that exists is itself complete.

I think this is a useful estimate; we should just make sure that we're up-front about how the estimate is calculated and what the assumptions are.

Having said all that, I don't exactly know where the right place to put this bit of explanation might be.  It's clearly too long for argument help text :)

timothytrippel replied:
Good point, best to clearly state this. Maybe I could add a short description in the legend beneath the table definition, next to the "Green = ..." and "Red = ..." (for reference see the updated table look below)? E.g.:
"Green = DIF API, as defined in the current header file, is complete. Note: the header file may lack necessary API functionality."
"Red = DIF API is incomplete, as defined in the header file."

Lmk what you think.

<img width="1164" alt="Screen Shot 2021-07-28 at 4 08 27 PM" src="https://user-images.githubusercontent.com/5633066/127407577-f14f05b3-72b0-403c-be27-dc089219a466.png">

cfrantz replied:
sgtm

timothytrippel replied:
added the short disclaimers to the table legend

asb commented:
> In addition to completing the above checklist items, I also:
> 
>     * verified authors, not committers, are indeed listed (the python module I am using, git-fame, uses the `git blame` command under the hood which make this distinction).
> 
>     * removed the "completed?" column, as this information is redundant given the color coding scheme/legend
> 
>     * added multi-line cells to table for width reduction and better displaying in the terminal
> 
> 
> One thing to note: when I was grabbing the DIF lifecycle states from the .prj.hjson files, if the IP had multiple revisions, I simply grabbed the latest.

This seems like a really useful tool, and I think given the constraint of pulling metadata just from the repository is doing a good job of presenting information. In terms of future development, we might think about whether it would be useful to add more metadata to the repository (perhaps reflecting what is tracked in external spreadsheets / project planning software). For instance, knowing the authors/contributors via git blame is useful, but you could also imagine adding something to prj.hjson to track the current "maintainer".

Perhaps bikeshedding a bit, but maybe for both "DIF Author" and "HW Author" a more neutral term might be "Contributors" (which feels like it more naturally includes the possibility of someone running up lots of LOC changes through refactoring).

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,311 @@
+    # Print the table and legend.
+    print("DIF Statuses:")
+    print(tabulate(rows, headers, tablefmt=TABLE_FORMAT))
+    print("*Only the lead author (by LOC) for each component is listed.")
```
Fixed.

timothytrippel commented:
> > In addition to completing the above checklist items, I also:
> > ```
> > * verified authors, not committers, are indeed listed (the python module I am using, git-fame, uses the `git blame` command under the hood which make this distinction).
> > 
> > * removed the "completed?" column, as this information is redundant given the color coding scheme/legend
> > 
> > * added multi-line cells to table for width reduction and better displaying in the terminal
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > One thing to note: when I was grabbing the DIF lifecycle states from the .prj.hjson files, if the IP had multiple revisions, I simply grabbed the latest.
> 
> This seems like a really useful tool, and I think given the constraint of pulling metadata just from the repository is doing a good job of presenting information. In terms of future development, we might think about whether it would be useful to add more metadata to the repository (perhaps reflecting what is tracked in external spreadsheets / project planning software). For instance, knowing the authors/contributors via git blame is useful, but you could also imagine adding something to prj.hjson to track the current "maintainer".
> 

Agreed. This would reduce ambiguity around who is actively responsible for various components (especially if someone leaves the project).

Additionally it may be useful for this tool include another flag that quickly returns the git log between when the hardware was last updated to when the software (DIF) was, so one does not have to manually cross-ref the dates listed in the table with the commit log. 

> Perhaps bikeshedding a bit, but maybe for both "DIF Author" and "HW Author" a more neutral term might be "Contributors" (which feels like it more naturally includes the possibility of someone running up lots of LOC changes through refactoring).

Fixed.

alphan reviewed (approved):
Thanks @timothytrippel! Just some minor comments, none of them are blocking.

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,314 @@
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Usage:
```
Fixed

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,314 @@
+    def __init__(self, repo_top, difs_root_path, dif_name):
+        # Get DIF/IP names and path.
+        self.dif_name = dif_name
+        self.ip = self.dif_name[4:]
```
Fixed.

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,314 @@
+        self.ip = self.dif_name[4:]
+        if self.ip in IP_TRANSLATION:
+            self.ip = IP_TRANSLATION[self.ip]
+        self.dif_path = os.path.join(difs_root_path, dif_name)
```
Maybe I'm missing something, but are you suggesting replacing all os.path.join's with pathlib calls? (If so I can definitely make that adjustment.) Or are you saying there is a way for pathlib to resolve partial matches in a potential path? Because the issue here is that for two of the IP blocks (namely, the rv_plic and entropy_src) the IP block names are different from the DIF names (e.g., "dif_plic.*" is the name for the DIF that actuates the "rv_plic" IP block)? If its the latter, that awesome! Could you elaborate more on this? (Admittedly, the way its written now is somewhat of a hack, but the IP block names deviate slightly from the DIF names in some cases, which probably should not be?)

alphan replied:
> are you suggesting replacing all os.path.join's with pathlib calls?

Not really, unless you really want to. The context that github shows is confusing but my comment was only for `os.path` usage in general. I think `pathlib` is cleaner and just wanted to mention that in case you are not familiar with it. For example,
```
os.path.join(self.hw_path, "data", self.ip + ".prj.hjson")
os.path.isdir(full_path[:-2])
```
would be something like
```
(self.hw_path / "data" / self.ip).with_suffix(".prj.hjson")
full_path.with_suffix('').is_dir()
```

> Or are you saying there is a way for pathlib to resolve partial matches in a potential path?

I'm not sure if it has something that you can use directly here, you could probably use `iterdir()` but wouldn't be cleaner than this.

timothytrippel replied:
Ah got it. Thanks!

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,314 @@
+    difs = get_list_of_difs(difs_root_path, shared_headers)
+
+    # Get DIF statuses (while displaying progress bar).
+    dif_statuses = []
+    print("Analyzing statuses of %s DIFs ..." % len(difs))
+    progress_bar_width = len(difs)
+    sys.stdout.write("[%s]" % (" " * progress_bar_width))
+    sys.stdout.flush()
+    sys.stdout.write(
+        "\b" * (progress_bar_width + 1))  # return to start of line, after '['
+    for dif in difs:
+        # Process a DIF.
+        dif_statuses.append(DIFStatus(args.repo_top, difs_root_path, dif))
+        # Update the progress bar.
+        sys.stdout.write("#")
+        sys.stdout.flush()
+    sys.stdout.write("]\n")  # end the progress bar
```
I originally chose a manual progress bar instead of tqdm b/c I didn't want to bloat the python-requirements.txt file but it seems tqdm is already a dependency of another packages so no sense not to use it. Nice catch!

sriyerg reviewed (commented):
This looks great, thanks! Could we add a shortcode to generate the table inline in the https://docs.opentitan.org/sw/ area?

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,314 @@
+            self.api_complete = False
+        # API is is incomplete --> SOME functions have not been implemented.
+        else:
+            self.api_complete = False
```
Fixed.

timothytrippel commented on util/check_dif_statuses.py:
```diff
@@ -0,0 +1,310 @@
+from termcolor import colored
+from tqdm import tqdm
+
+# The IP suffix of some DIFs are slightly different than the IP directory name
+# in the hw/ip/ directory, so here is the translation for the applicable IPs.
+IP_TRANSLATION = {
+    "plic": "rv_plic",
+    "entropy": "entropy_src",
+}
```
> Is there a good reason for this deviation?
> `rv_timer` retains the `rv_` prefix. Plus, we loosely associate 'entropy' with the whole entropy SS rather than with the `entropy_src` IP. 

Probably not, but other should weigh in on this? 

> IMO it would be better to have the DIF names match the IP names.

Agreed, according to DIF style guide names should match the IP block, so weird that this ended up not matching. We should address the DIF name / IP name mismatch in separate PR.

timothytrippel commented:
> This looks great, thanks! Could we add a shortcode to generate the table inline in the https://docs.opentitan.org/sw/ area?

Yes! I can definitely do that (I'm using the tabulate module to generate the table so we can output several formats, Markdown being one). However, would [this](https://docs.opentitan.org/sw/device/lib/dif/README/) would be a better location?

timothytrippel commented:
> > This looks great, thanks! Could we add a shortcode to generate the table inline in the https://docs.opentitan.org/sw/ area?
> 
> Yes! I can definitely do that (I'm using the tabulate module to generate the table so we can output several formats, Markdown being one). However, would [this](https://docs.opentitan.org/sw/device/lib/dif/README/) would be a better location?

Added arg. to specify table format. Choices are a subset of those documented by the Python [tabulate](https://pypi.org/project/tabulate/) module. All of which are forms of Markdown.

timothytrippel commented:
@cfrantz Added type annotations and docstrings.

timothytrippel commented:
If there are no more comments/concerns could we merge this? Thanks!

sriyerg commented on python-requirements.txt:
```diff
@@ -29,6 +31,8 @@ pytest
 pyyaml
 tabulate
+termcolor
+tqdm
```
Could you use `enlighten` instead, which does the same thing as `tqdm`? Would be good to be able to minimize external dependencies.

timothytrippel replied:
Fixed.

sriyerg commented:
> > This looks great, thanks! Could we add a shortcode to generate the table inline in the https://docs.opentitan.org/sw/ area?
> 
> Yes! I can definitely do that (I'm using the tabulate module to generate the table so we can output several formats, Markdown being one). However, would [this](https://docs.opentitan.org/sw/device/lib/dif/README/) would be a better location?

Yes, that's indeed a better location. Thanks!

timothytrippel commented:
> > > This looks great, thanks! Could we add a shortcode to generate the table inline in the https://docs.opentitan.org/sw/ area?
> > 
> > 
> > Yes! I can definitely do that (I'm using the tabulate module to generate the table so we can output several formats, Markdown being one). However, would [this](https://docs.opentitan.org/sw/device/lib/dif/README/) would be a better location?
> 
> Yes, that's indeed a better location. Thanks!

Script argument (--table-format) added. Will address adding to table to website in a separate PR.

sriyerg reviewed (approved):
LGTM, thanks!

alphan commented:
Merging since this PR received multiple approvals and all review feedback has been addressed.
